### PR TITLE
Fix TLS verification

### PR DESCRIPTION
**Pull request by marusama**
Fix TLS verification via hostname

### CODE REVIEW

**Review by aglarendil — APPROVED** [review]

File: rediscluster/cluster.go [hunk 1]
```diff
@@ -95,8 +95,12 @@ type Opts struct {
...
 	// LatencyOrientedRR - when MasterAndSlaves is used, prefer hosts with lower latency
 	LatencyOrientedRR bool
-	TLSEnabled        bool
-	TLSConfig         *tls.Config
+	// Enable connection with TLS
+	TLSEnabled bool
+	// Config for TLS connection
+	TLSConfig *tls.Config
+	// Do not resolve host address to ip
+	SkipHostResolving bool
```
Comment (g7r) [inline rediscluster/cluster.go:103]: It looks like there is no need in `SkipHostResolving`.

**Review by g7r — CHANGES_REQUESTED** [review]

File: rediscluster/cluster.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ const (
 	preferConnected
 )
 
-// Opts is a options for Cluster
+// Opts is an options for Cluster
```
[the remark this answers is not in the document]
Comment (marusama) [reply rediscluster/cluster.go]: changed

File: rediscluster/cluster.go (hunk 1 above)
Comment (marusama) [reply rediscluster/cluster.go]: `SkipHostResolving` removed

File: rediscluster/mapping.go [hunk 3]
```diff
@@ -57,6 +73,33 @@ func (c *Cluster) newNode(addr string, initial bool) (*node, error) {
...
+func (c *Cluster) nodeOpts(addr string) (*redisconn.Opts, error) {
+	nodeOpts := c.opts.HostOpts
+
+	if !nodeOpts.TLSEnabled {
+		return &nodeOpts, nil
+	}
+
+	originalHost, err := redisclusterutil.GetHost(addr)
+	if err != nil {
+		return nil, err
```
Comment (g7r) [inline rediscluster/mapping.go:85]: The problem here is that we return a naked error from `net.SplitHostPort`. As you may have noticed, `redispipe` is very careful when it comes to dealing with errors. I believe we should wrap the error here in a package-specific error type, even though it's almost impossible to have an error here.

File: rediscluster/mapping.go [hunk 4]
```diff
@@ -57,6 +73,33 @@ func (c *Cluster) newNode(addr string, initial bool) (*node, error) {
...
+	}
+
+	if !redisclusterutil.IsIPAddress(originalHost) {
+		// preserve original hostname for TLS verification
+		if nodeOpts.TLSConfig != nil {
+			nodeOpts.TLSConfig = nodeOpts.TLSConfig.Clone()
+			nodeOpts.TLSConfig.ServerName = originalHost
```
Comment (g7r) [inline rediscluster/mapping.go:92]: This is a common code between the two branches, but it doesn't seem that way. I think we should move it outside of this `if`.

**Review by g7r — CHANGES_REQUESTED** [review]

File: rediscluster/mapping.go (hunk 3 above)
Comment (marusama) [reply rediscluster/mapping.go]: fixed

File: rediscluster/mapping.go (hunk 4 above)
Comment (marusama) [reply rediscluster/mapping.go]: fixed

**Review by g7r — APPROVED** [review]

**Pull request merged**